torch/distributed/_composable/replicate.py

```diff
@@ -0,0 +1,107 @@
+from typing import List, Tuple
+
+import torch
+import torch.nn as nn
+
+from . import _ddp
+from .contract import contract
+
+
+class DistributedState:
+    ...
+
+
+class ReplicateState(DistributedState):
+    def __init__(self) -> None:
+        self.modules: List[nn.Module] = []
+        self.has_initialized: bool = False
+        self._param_list: nn.ParameterList = nn.ParameterList()
+
+    def mark_modules(self, *modules: nn.Module) -> None:
+        for module in modules:
+            self.modules.append(module)
+            replicate.state(module)._distributed_state = self
+            replicate.state(module)._params_collected = False
+
+    def _recursive_collect_params(self, module: nn.Module) -> None:
+        # TODO: skip if managed by other APIs
+
+        if hasattr(replicate.state(module), "_params_collected"):
+            if replicate.state(module)._params_collected:
+                return
+            replicate.state(module)._params_collected = True
+
+        self._param_list.extend(
+            param
+            for param in module.parameters(recurse=False)
+            # for param in module.parameters()
+            if param.requires_grad
+        )
+        for child in module.children():
+            self._recursive_collect_params(child)
+
+    def init_helper(self):
+        if self.has_initialized:
+            return
+
+        self.has_initialized = True
+        for module in self.modules:
+            self._recursive_collect_params(module)
+
+        self._ddp = _ddp.DistributedDataParallel(self._param_list)
+
+    def root_module_forward_pre_hook(
+        self, module: nn.Module, input: Tuple[torch.Tensor]
+    ) -> None:
+        self.init_helper()
+        self._ddp.pre_forward()
```

Comment:
same as above, we need to make pre_forward() accept a replicate.state() object once get rid of self._ddp?